### PR TITLE
Initial commit of custom-elements-scoped proof-of-concept

### PR DESCRIPTION
**Pull request by kevinpschaaf**
Initial commit of scoped `CustomElementRegistry` proof-of-concept polyfill, to open up community collaboration on moving the polyfill forward (continuation of work done in this [gist](https://gist.github.com/kevinpschaaf/dc80e70adabb5103cbfacae75be33471)).

Big tasks needed to move this forward (in subsequent PR's):
* [x] Add test suite
* [ ] Convert source to TS (to match convention in this monorepo)
* [ ] Test and work out layering strategy with standard `custom-elements` and `shady-dom` polyfills
* [ ] Consider adding benchmarks
===

Edit: remaining TODO's moved to issues:
* [#419](https://github.com/webcomponents/polyfills/issues/419): Convert source to TS (to match convention in this monorepo)
* [#420](https://github.com/webcomponents/polyfills/issues/420): Test and work out layering strategy with standard custom-elements polyfill
* [#421](https://github.com/webcomponents/polyfills/issues/421): Test and work out layering strategy with shady-dom & shady-css polyfills
* [#422](https://github.com/webcomponents/polyfills/issues/422): Add benchmarks

### CODE REVIEW

File: packages/custom-elements-scoped/package.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "@webcomponents/custom-elements-scoped",
```
Comment (aomarks) [inline packages/custom-elements-scoped/package.json:2]: Maybe `scoped-custom-elements`? (For some reason `custom-elements-scoped` to me kinda sounds like it's a variant of the full custom-elements polyfill that also supports scoped). But I don't have a strong opinion.

**Review by aomarks — APPROVED** [review]
Really cool implementation!

File: packages/custom-elements-scoped/custom-elements-scoped.js [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+window.CustomElementRegistry = class {
+  constructor(options) {
+    this._definitions = new Map();
+    this._definedPromises = new Map();
+    this._definedResolvers = new Map();
+    this._awaitingUpgrade = new Map();
+    // Once the registry is patched, this is easy to allow (previously upgraded
+    // elements won't be affected, but new elements will be upgraded with the
+    // new definition)
+    this._allowRedefinition = options && options.allowRedefinition;
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/custom-elements-scoped/custom-elements-scoped.js:37]: It was a little gratuitous addition I made while experimenting; basically the implementation makes it easy to allow last-in re-definition. I should probably pull it out, good catch.

File: packages/custom-elements-scoped/custom-elements-scoped.js [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    // are simulated via setAttribute/removeAttribute patches
+  };
+}
+
+// Helper to patch CE class setAttribute/getAttribute to implement
+// attributeChangedCallback
+const patchAttributes = (elementClass, observedAttributes, attributeChangedCallback) => {
+  if (observedAttributes.size && attributeChangedCallback) {
+    const setAttribute = elementClass.prototype.setAttribute;
+    if (setAttribute) {
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/custom-elements-scoped/custom-elements-scoped.js:217]: Not possible, it's on `HTMLElement.prototype`.

File: packages/custom-elements-scoped/package.json (hunk 1 above)
Comment (kevinpschaaf) [reply packages/custom-elements-scoped/package.json:2]: 👍 Yeah, I take your point. I'm not sure flipping it helps a ton with that particular confusion (i.e. do I need both `custom-elements` and `scoped-custom-elements`?). But at a minimum it reads less awkward that way.

Comment (LarsDenBakker) [reply packages/custom-elements-scoped/package.json:2]: Should we have registry in the name?

`@webcomponents/scoped-registry`
`@webcomponents/scoped-elements-registry`
`@webcomponents/scoped-custom-elements-registry`

File: packages/custom-elements-scoped/README.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+## 🚨 Work in progress
+
+This polyfill explores implementation of a non-finalized spec proposal and is
+currently a work in progress.
+
+## Overview
+
+Scoped CustomElementRegistry polyfill based on discussion in
+https://github.com/w3c/webcomponents/issues/716 and spec proposal in 
+https://github.com/w3c/webcomponents/pull/865.
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/custom-elements-scoped/README.md:12]: Done.

File: packages/custom-elements-scoped/custom-elements-scoped.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/custom-elements-scoped/custom-elements-scoped.js:37]: Removed, per below.

[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/custom-elements-scoped/custom-elements-scoped.js:37]: Removed.

File: packages/custom-elements-scoped/custom-elements-scoped.js [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    }
+    // no attributeChangedCallback or observedAttributes since these
+    // are simulated via setAttribute/removeAttribute patches
+  };
+}
+
+// Helper to patch CE class setAttribute/getAttribute to implement
+// attributeChangedCallback
+const patchAttributes = (elementClass, observedAttributes, attributeChangedCallback) => {
+  if (observedAttributes.size && attributeChangedCallback) {
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/custom-elements-scoped/custom-elements-scoped.js]: Done

File: packages/custom-elements-scoped/custom-elements-scoped.js [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+    const attributeChangedCallback = elementClass.prototype.attributeChangedCallback;
+    const observedAttributes = new Set(elementClass.observedAttributes || []);
+    patchAttributes(elementClass, observedAttributes, attributeChangedCallback);
+    // Register the definition
+    const definition = {
+      elementClass,
+      connectedCallback: elementClass.prototype.connectedCallback,
+      disconnectedCallback: elementClass.prototype.disconnectedCallback,
+      adoptedCallback: elementClass.prototype.adoptedCallback,
+      attributeChangedCallback: attributeChangedCallback,
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/custom-elements-scoped/custom-elements-scoped.js]: Done

File: packages/custom-elements-scoped/custom-elements-scoped.js [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+};
+
+// Helper to create stand-in element for each tagName registered that delegates
+// out to the registry for the given element
+const createStandInElement = (tagName) => {
+  return class ScopedCustomElementBase {
+    constructor() {
+      // Create a raw HTMLElement first
+      const instance = Reflect.construct(NativeHTMLElement, [], this.constructor);
+      // We need to install the minimum the HTMLElement prototype so that
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/custom-elements-scoped/custom-elements-scoped.js]: Done

File: packages/custom-elements-scoped/custom-elements-scoped.js [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+const scopeForNode = (node) => {
+  // TODO: this is not per the proposal; assigning a one-time scope at creation
+  // time would require walking every tree ever created, which is avoided for now
+  let scope = node.getRootNode();
+  // If not, get the tree scope from the context that created the node;
+  // if that is not in a tree root, then bail
+  if (!isValidScope(scope)) {
+    scope = creationContext[creationContext.length-1].getRootNode();
+    if (!isValidScope(scope)) {
+      throw new Error('Element is being upgrade outside of a valid tree scope');
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/custom-elements-scoped/custom-elements-scoped.js]: Done

File: packages/custom-elements-scoped/custom-elements-scoped.js [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+    }
+    if (this === window.customElements) {
+      globalDefinitionForConstructor.set(elementClass, definition);
+      definition.standInClass = standInClass;
+    }
+    // Upgrade any elements created in this scope before define was called
+    const awaiting = this._awaitingUpgrade.get(tagName);
+    if (awaiting) {
+      this._awaitingUpgrade.delete(tagName);
+      awaiting.forEach(element => {
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/custom-elements-scoped/custom-elements-scoped.js]: Done

File: packages/custom-elements-scoped/custom-elements-scoped.js [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+    // Flush whenDefined callbacks
+    const resolver = this._definedResolvers.get(tagName);
+    if (resolver) {
+      resolver();
+    }
+    return elementClass;
+  }
+  get(tagName) {
+    const definition = this._definitions.get(tagName);
+    return definition ? definition.elementClass : undefined;
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/custom-elements-scoped/custom-elements-scoped.js]: Done

File: packages/custom-elements-scoped/custom-elements-scoped.js [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+        // Delegate out to user callback
+        definition.disconnectedCallback && definition.disconnectedCallback.apply(this, arguments);
+      } else {
+        // Un-register for upgrade when defined (so we don't leak)
+        pendingRegistryForElement.get(this)._upgradeWhenDefined(this, tagName, false);        
+      }
+    }
+    adoptedCallback() {
+      const definition = definitionForElement.get(this);
+      if (definition && definition.adoptedCallback) {
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/custom-elements-scoped/custom-elements-scoped.js]: Done

File: packages/custom-elements-scoped/package.json (hunk 1 above)
Comment (kevinpschaaf) [reply packages/custom-elements-scoped/package.json]: Changed to `scoped-custom-elements-registry`

**Comment by kevinpschaaf** [comment]
@manolakis Sorry for the delay, thanks so much for the work on this! I got the tests passing now.

Could you add the ["googlebot I consent"](https://github.com/webcomponents/polyfills/pull/383#issuecomment-742364017) comment, and also review the changes I made in this commit: https://github.com/webcomponents/polyfills/pull/383/commits/825dc926780ef6d640a111d7410adb812bc1d70e, particularly the change I made to this test, which I _think_ was a mistake in the test, but I may heave mis-understood the intention: https://github.com/webcomponents/polyfills/pull/383/commits/825dc926780ef6d640a111d7410adb812bc1d70e#diff-b6d5700c31b8bc7d7578717fa944702bb7441e04cfadeeb426fe04db42e45a1dR42

**Comment by kevinpschaaf** [comment]
The algorithm for determining the registry during upgrade is pretty shaky, and I think feedback from this is gonna need to be fed back into the spec proposal.  Particularly this case:
```js
const registry = new CustomElementRegistry();
registry.define('x-foo', class extends HTMLElement{});
const shadowRoot = document.body.attachShadow({mode: 'open', registry});
const div = shadowRoot.createElement('div');
div.innerHTML = `<x-foo></x-foo>`;
```
Without having stored state on the `div` noting its creation tree scope, the `innerHTML` (which the proposal was trying to avoid) I don't see a good way for the `innerHTML` to know to use the shadow root's registry. I did this to get the test to pass (via the `scopeForNode` `WeakMap`), but I only store this for nodes directly created by `createElement`; if you had `innerHTML`'ed some DOM into that div, and then did a further `innerHTML` into one of _those_ nodes, it would fail, since I'm trying to avoid needing to treewalk every node ever created. cc @justinfagnani

**Comment by manolakis** [comment]
@googlebot I consent

**Comment by manolakis** [comment]
@kevinpschaaf I agree that the test is a bit confusing.

My intention was to check what should happen if there is a node in the light DOM which contains an upgraded WC defined in the global registry, and we want to import it into a shadowDom with a custom registry. What should be the behaviour in this case? Should the instance of the component be of the class defined in the global registry or should be a non upgraded component becase that tag name is not defined in the custom registry?

**Comment by leobalter** [comment]
@kevinpschaaf I'd like to support with some traction here. Are the OP checkboxes up to date? Is there anything specific to take a look?

**Comment by kevinpschaaf** [comment]
@leobalter Sorry, haven't been able to focus on this lately. I think this PR is basically ready to land, save for figuring out what we want to do for the test I raised with @manolakis, which I hadn't had a chance to circle back yet.

And then yeah, I think the next things would be from that checklist. Switching the code to TS is more of a chore sort of thing, not necessarily critical path but would be nice. Working out polyfill compatibility is probably the more important chunk of work, assuming that's important to your use cases. I expect custom-elements to be fairly straightforward, but ShadyDOM (ShadyCSS) is going to be pretty rough, since right now it scopes element CSS by tag name; the ability to have multiple tag names scoped differently would require adding an additional scoping class/attribute per registry.  So I guess a question is... how important is the legacy polyfill story to your use of the scoped registry polyfill, and are you interested in looking at those integrations?

**Comment by kevinpschaaf** [comment]
@manolakis Re: the test we were discussing, if we're talking about importing a node, my expectation is that should be purely a function of the node's tagName/attributes and the new context it is being imported into (i.e. whether or not it had a custom element identity before it was imported would be immaterial)... So I could change the test back but make the expectation to be an un-upgrade `HTMLElement`?  Or else we just leave the test as-is and it just tests a different behavior? Wdyt?

**Comment by manolakis** [comment]
@kevinpschaaf Yes, you are right. I tried to clarify the test intention in this PR against this branch https://github.com/webcomponents/polyfills/pull/412. All test green :+1:

**Comment by kevinpschaaf** [comment]
@manolakis Right! Sorry, that LGTM, I merged that in. And I'll go ahead and merge this into master, and we can take new PR's from there.

**Comment by kevinpschaaf** [comment]
@leobalter For sake of merging this PR, I moved the outstanding TODO's to issues and noted them in the README:

* [#419](https://github.com/webcomponents/polyfills/issues/419): Convert source to TS (to match convention in this monorepo)
* [#420](https://github.com/webcomponents/polyfills/issues/420): Test and work out layering strategy with standard custom-elements polyfill
* [#421](https://github.com/webcomponents/polyfills/issues/421): Test and work out layering strategy with shady-dom & shady-css polyfills
* [#422](https://github.com/webcomponents/polyfills/issues/422): Add benchmarks

**Pull request merged**